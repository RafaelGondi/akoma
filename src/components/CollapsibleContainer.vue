<template>
	<span id="collapsible-container">
		<div
			class="collapsible-container__item"
			@click="triggerExpanded"
		>
			<span
				:class="internalValue ? 'collapsible-container__chevron--opened' : 'collapsible-container__chevron--closed'"
			/>
			<!-- @slot Slot para renderização do conteúdo do header/title customizado do item
			-->
			<slot name="title">
				<span
					class="collapsible-container__title"
				>
					{{ title }}
				</span>
			</slot>
		</div>
		<div
			v-if="internalValue"
			class="collapsible-container__content"
		>
			<!-- @slot Slot para renderização conteúdo do container
				quando o componente está expandido
			-->
			<slot />
		</div>
	</span>
</template>

<script>
export default {
	props: {
		/**
		 * Representa o estado do componente. Quando 'false' o conteúdo não é mostrado
		 * e quando 'true' o conteúdo do container é exibido.
		 */
		modelValue: {
			type: Boolean,
			default: false,
			required: true,
		},
		/**
		 * Título do item a ser expandido
		 */
		title: {
			type: String,
			default: 'Visualizar mais',
		},
	},

	data() {
		return {
			internalValue: this.modelValue,
		};
	},

	watch: {
		internalValue(newValue) {
			/**
			* Evento emitido quando o conteúdo do item é exibido ('true') ou ocultado ('false').
			* @event expanded
			* @type {Event}
			*/
			this.$emit('update:modelValue', newValue);
		},

		modelValue(newValue) {
			this.internalValue = newValue;
		},
	},

	methods: {
		triggerExpanded() {
			this.internalValue = !this.internalValue;
		},
	},
};
</script>

<style lang="scss">
@import '../assets/sass/tokens.scss';

.collapsible-container {
	&__content {
		padding: px(5);

		& p {
			margin: my(1);
		}
	}

	&__chevron--closed {
		top: 6px;
		right: 2px;
		display: block;
		height: 32px;
		width: 32px;
		pointer-events: none;
		border-radius: $border-radius-small;
		display: flex;
		align-items: center;
		justify-content: center;

		&::before,
		&::after {
			content: '';
			width: 0;
			height: 0;
		}

		&::before {
			border-left: 6px solid transparent;
			border-right: 6px solid transparent;
			border-bottom: 6px solid $n-400;
			border-radius: $border-radius-small;
			transition: all 300ms ease-in-out;
			transform: rotate(-180deg);
		}
	}

	&__chevron--opened {
		top: 6px;
		right: 2px;
		display: block;
		height: 32px;
		width: 32px;
		pointer-events: none;
		border-radius: $border-radius-small;
		display: flex;
		align-items: center;
		justify-content: center;

		&::before,
		&::after {
			content: '';
			width: 0;
			height: 0;
		}

		&::before {
			border-left: 6px solid transparent;
			border-right: 6px solid transparent;
			border-bottom: 6px solid $n-400;
			border-radius: $border-radius-small;
			transition: all 300ms ease-in-out;
			transform: rotate(0deg);
		}
	}

	&__item {
		font-size: 14px;
		color: $n-600;
		display: flex;
		align-items: start;
		cursor: pointer;
		display: flex;
		align-items: baseline;
		margin: ml(n2);

		& hr {
			border-color: $n-600 !important;
		}
	}

	&__title {
		font-weight: 700;
	}

	&__icon--collapsed {
		color: $n-600;
		margin: mr(2);
		transition: all 0.25s ease-in-out;
	}

	&__icon--expanded {
		@extend .collapsible-container__icon--collapsed;
		transform: rotate(180deg);
	}
}
</style>
