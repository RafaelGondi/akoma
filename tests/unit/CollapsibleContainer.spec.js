import { mount, shallowMount, createLocalVue } from '@vue/test-utils';
import { BootstrapVue, BootstrapVueIcons } from 'bootstrap-vue';
import CollapsibleContainer from '../../src/components/CollapsibleContainer.vue';
import flushPromises from 'flush-promises';

const localVue = createLocalVue();

localVue.use(BootstrapVue);
localVue.use(BootstrapVueIcons);

test('Component is mounted properly', () => {
	const wrapper = shallowMount(CollapsibleContainer, {
		localVue,
		propsData: {
			value: true,
			title: 'Seu universo expandido',
		},
		stubs: {
			'ion-icon': true,
		},
	});

	expect(wrapper).toMatchSnapshot();
});

describe('Click tests', () => {
	test('if container is expanded and an event is emitted when the item is clicked', async () => {
		const wrapper = shallowMount(CollapsibleContainer, {
			localVue,
			propsData: {
				title: 'Expanda seu universo',
				value: false,
			},
			stubs: {
				'ion-icon': true,
			},
		});

		wrapper.find('.cds-collapsible-container__item').trigger('click');
		await flushPromises();

		expect(wrapper.find('.cds-collapsible-container__content').exists()).toBe(true);
		expect(wrapper.find('.cds-collapsible-container__icon--expanded').exists()).toBe(true);

		expect(wrapper.emitted().input).toBeTruthy();
		expect(wrapper.emitted().input).toEqual([[true]]);
	});

	test('if container is collapsed and an event is emitted when the item is clicked', async () => {
		const wrapper = shallowMount(CollapsibleContainer, {
			localVue,
			propsData: {
				title: 'Includa seu universo',
				value: true,
			},
			stubs: {
				'ion-icon': true,
			},
		});

		wrapper.find('.cds-collapsible-container__item').trigger('click');
		await flushPromises();

		expect(wrapper.find('.cds-collapsible-container__content').exists()).toBe(false);
		expect(wrapper.find('.cds-collapsible-container__icon--collapsed').exists()).toBe(true);

		expect(wrapper.emitted().input).toBeTruthy();
		expect(wrapper.emitted().input).toEqual([[false]]);
	});
});
