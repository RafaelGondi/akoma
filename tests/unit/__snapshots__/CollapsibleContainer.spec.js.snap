// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component is mounted properly 1`] = `
<span id="cds-collapsible-container"><div class="cds-collapsible-container__item"><ion-icon-stub name="chevron-down-outline" class="cds-collapsible-container__icon--expanded"></ion-icon-stub> <span class="cds-collapsible-container__title">
				Seu universo expandido
			</span></div>
<div class="cds-collapsible-container__content"></div>
</span>
`;
